lib/active_record/virtual_attributes/virtual_total.rb

```diff
@@ -98,31 +98,36 @@ def define_virtual_aggregate_method(name, relation, method_name, column)
       end
 
       def virtual_aggregate_arel(reflection, method_name, column)
-        return unless reflection && reflection.macro == :has_many && !reflection.options[:through]
+        return unless reflection && reflection.macro == :has_many
+
+        # need db access for the reflection join_keys, so delaying all this key lookup until call time
         lambda do |t|
-          query = if reflection.scope
-                    reflection.klass.instance_exec(nil, &reflection.scope)
-                  else
-                    reflection.klass.all
-                  end
-
-          foreign_table = reflection.klass.arel_table
-          # need db access for the keys, so delaying all this lookup until call time
-          if ActiveRecord.version.to_s >= "5.1"
-            join_keys = reflection.join_keys
-          else
-            join_keys = reflection.join_keys(reflection.klass)
-          end
-          query       = query.except(:order).where(t[join_keys.foreign_key].eq(foreign_table[join_keys.key]))
-
-          arel_column = if method_name == :size
-                          Arel.star.count
-                        else
-                          reflection.klass.arel_attribute(column).send(method_name)
-                        end
-          query       = query.select(arel_column)
-
-          t.grouping(Arel::Nodes::SqlLiteral.new(query.to_sql))
+          # arel_column: COUNT(*)
+          arel_column = (column == Arel.star ? column : reflection.klass.arel_attribute(column)).send(method_name)
```

Comment:
I wanted to call arel_attribute if column is a string or symbol, but `Arel.star.kind_of?(String)`

The goal of using `Arel.star` is to avoid the hardcoded logic here. (especially basing the column upon the method field)

skipping the `if` check essentially calls `reflection.klass.arel_attribute(Arel.star)` - which will call `count(table.*)` which is good on pg but fails in sqlite and mysql - and is not really what I was aiming for.

I'm very happy how passing `method_name` as `:count` works now. much better than passing `:size` and changing to `:count` here.

BUT this change is not necessary (it is the last commit) and I could remove it from this PR.
I just through since we are focusing on testing the counts here, it would make sense to have both changes together.

Reply:
fixed